### PR TITLE
[https://nvbugs/5392414] [fix] Add customized default routing method

### PR DESCRIPTION
**Pull request by ChristinaZ**
## Summary by CodeRabbit

- **New Features**
  - Faster MOE Top‑K path for small inputs (len ≤128, k ≤8); BF16 support and optional softmax-before-topk routing exposed.
  - New default MOE routing Torch operator and Python flag (force_enable_pytorch_op) to force a PyTorch fallback.

- **Changes**
  - Top‑K dispatch now routes eligible small inputs through the MOE path; iterators unified to Thrust and index type standardized.
  - Runtime dispatch and launcher updated to select specialized MOE kernels for small cases.

- **Removals**
  - Legacy renorm MOE routing kernel and its old launcher removed.

## Defauled routing-related optimization
1. For the Pytorch backend, added a customized Pytorch Op.
2. For the TRT backend, add one more customized topK kernel special for the MoE topK calculation. Now it only takes 3 us for the sample from the customer.
image

## Test Coverage
Follow the example in nvBug.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: cpp/tensorrt_llm/thop/defaultMoeRoutingOp.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "tensorrt_llm/runtime/torchUtils.h"
+
+namespace th = torch;
+namespace tl = tensorrt_llm;
+namespace tk = tensorrt_llm::kernels;
+
+namespace torch_ext
+{
+
+std::tuple<at::Tensor, at::Tensor> default_moe_routing_op(th::Tensor const& router_logits, int64_t topk)
```
Comment (djns99) [inline cpp/tensorrt_llm/thop/defaultMoeRoutingOp.cpp:28]: Can we make this use the same function as renorm so we don't have to update both places

File: cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu [hunk 2]
```diff
@@ -306,19 +189,27 @@ int nextPowerOfTwo(int num)
     return power;
 }
 
-#define CASE(MAX_NUM_EXPERTS)                                                                                          \
+#define CASE(MAX_NUM_EXPERTS, DoSoftmaxBeforeTopK)                                                                     \
```
Comment (djns99) [inline cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu:192]: This parameter isn't needed, you can just use it directly like InputT etc.

File: cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu [hunk 3]
```diff
@@ -253,36 +113,59 @@ __global__ void renormMoeRoutingKernel(InputT* routerLogits, OutputT* topkValues
...
-        auto score = calcSoftmax(warp,
-            laneIdx < topK ? static_cast<float>(warpTopKScore[laneIdx]) : static_cast<float>(minScore), laneIdx, topK);
-        if (laneIdx < topK)
+        // Normalize the scores
+        if constexpr (DoSoftmaxBeforeTopK)
         {
-            topkValues[outputOffset + laneIdx] = static_cast<OutputT>(score);
-            topkIndices[outputOffset + laneIdx] = warpTopKExpertIdx[laneIdx];
+            float sum = float{1.f};
+            if (normTopkProb)
```
Comment (djns99) [inline cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu:148]: Is this implementing the naive renormalize? If so I don't think we really need to support that here do we?

**Review by djns99 — APPROVED** [review]
LGTM, some potential style refactors but otherwise good

File: cpp/tensorrt_llm/thop/defaultMoeRoutingOp.cpp (hunk 1 above)
Comment (ChristinaZ) [reply cpp/tensorrt_llm/thop/defaultMoeRoutingOp.cpp:28]: Good point. Let me revise it.

File: cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu (hunk 2 above)
Comment (ChristinaZ) [reply cpp/tensorrt_llm/kernels/renormMoeRoutingKernels.cu:192]: Agree.

**Review by QiJune — APPROVED** [review]
LGTM

**Comment by QiJune** [comment]
@ChristinaZ  Please cherry pick this fix to release/1.0 branch as well.

**Comment by litaotju** [comment]
What performance tests had been done for this kernel optimization?

How it affects the DS R1 performance on Hopper and Blackwell? Thx


Also I didn't see any model file change after ading this new kernel, do we need that.

File: cpp/tensorrt_llm/kernels/trtllmGenKernels/blockScaleMoe/RoutingRenormalize.cu [hunk 4]
```diff
@@ -199,13 +199,11 @@ __global__ void __launch_bounds__(NumThreadsSingleBlock) routingIndicesBlockKern
     }
 
 #if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
-#if !defined(PDL_PROFILE) || PDL_PROFILE == 0
```
[the remark this answers is not in the document]
Comment (ChristinaZ) [reply cpp/tensorrt_llm/kernels/trtllmGenKernels/blockScaleMoe/RoutingRenormalize.cu:202]: Oh, yes. This macro variable was intended for debugging and profiling. Previously its name violated the guardword in the post-merge pipeline. We revised it first with PR https://github.com/NVIDIA/TensorRT-LLM/pull/6983. Then after discussing June, since this macro variable is only used for debugging and profiling, we can remove it.

**Review by litaotju — APPROVED** [review]
